src/Layer/LabelLayer.js

```diff
@@ -0,0 +1,221 @@
+    }
+
+    /**
+     * Reads each {@link FeatureGeometry} that contains label configuration, and
+     * creates the corresponding {@link Label}. To create a `Label`, a geometry
+     * needs to have a `label` object with at least a few properties:
+     * - `content`, which refers to `Label#content`
+     * - `position`, which refers to `Label#position`
+     * - (optional) `config`, containing miscellaneous configuration for the
+     *   label
+     *
+     * The geometry (or its parent Feature) needs to have a Style set.
+     *
+     * @param {FeatureCollection} data - The FeatureCollection to read the
+     * labels from.
+     * @param {Extent} extent
+     *
+     * @return {Label[]} An array containing all the created labels.
+     */
+    convert(data, extent) {
+        const labels = [];
+
+        const layerField = this.style && this.style.text && this.style.text.field;
+
+        // Converting the extent now is faster for further operation
+        extent.as(data.crs, _extent);
+        coord.crs = data.crs;
+
+        data.features.forEach((f) => {
+            // TODO: add support for LINE and POLYGON
+            if (f.type !== FEATURE_TYPES.POINT) {
+                return;
+            }
+
+            const featureField = f.style && f.style.text.field;
+
+            f.geometry.forEach((g) => {
+                const minzoom = (g.properties.style && g.properties.style.zoom && g.properties.style.zoom.min)
+                    || (f.style && f.style.zoom && f.style.zoom.min)
+                    || (this.style && this.style.zoom && this.style.zoom.min);
+
+                // Don't create a label if it is in-between two steps of zoom
+                if (!this.source.isFileSource) {
+                    if (data.extent.zoom != minzoom) { return; }
+                } else if (extent.zoom != minzoom) { return; }
+
+                // NOTE: this only works because only POINT is supported, it
+                // needs more work for LINE and POLYGON
+                coord.setFromArray(f.vertices, g.size * g.indices[0].offset);
+                data.transformCoordinates(coord);
+                if (f.size == 2) { coord.z = 0; }
+                if (!_extent.isPointInside(coord)) { return; }
+
+                const geometryField = g.properties.style && g.properties.style.text.field;
+                if (!geometryField && !featureField && !layerField) {
+                    return;
+                } else if (geometryField) {
+                    content = g.properties.style.getTextFromProperties(g.properties);
+                } else if (featureField) {
+                    content = f.style.getTextFromProperties(g.properties);
+                } else if (layerField) {
+                    content = this.style.getTextFromProperties(g.properties);
+                }
+
+                // FIXME: Style management needs improvement, see #1318.
+                const label = new Label(content,
+                    coord.clone(),
```

Comment:
If is possible, could you remove `clone()`?

Reply:
Well `coord` is global, so I can't really avoid this.

Reply:
ok